### DISCUSSION TITLE
Defining Typed Collections that start empty

### DISCUSSION THREAD

**Discussion by jamespavett** [larastan/larastan#1140] [Q&A] [answered]
I raise this earlier today on Larastan and Laravel repo. Was instructed to raise it here instead: 

- Larastan Version: 2.0.1
- `--level` used: 6

### Description

Basically I am typehinting a Support Collection as containing `<int, DummyType>`

However in the constructor I want to initialise this to an empty collection, and then add `DummyType` elements to it late on. Is there a way to probably typehint this to Larastan / PHPStan doesn't throw errors.

Currently I am getting:

`Property App\Services\DummyType Service::$generatedData (Illuminate\Support\Collection<int, App\Services\DummyType>) does not accept Illuminate\Support\Collection<*NEVER*, *NEVER*>.`

Or (when initialising collection via collect helper):

`Unable to resolve the template type TKey in call to function collect`
`Unable to resolve the template type TValue in call to function collect`

Not sure if I am trying to do something wrong here, or there is another way to achieve what I am after

### Laravel code where the issue was found

```php
    /**
     * The retrieved query data
     *
     * @var \Illuminate\Support\Collection<int, object>
     */
    private Collection $queryData;

    /**
     * The generated combined data
     *
     * @var \Illuminate\Support\Collection<int, DummyType>
     */
    private Collection $generatedData;

    /**
     * Constructor.
     */
    public function __construct()
    {

        $this->generatedData = new Collection();

        // or

        $this->generatedData = collect();
    }
```

Just one other weird Thing I have noticed. If I change the typehint to (removing the int bit):

```php
    /**
     * The generated combined data
     *
     * @var \Illuminate\Support\Collection<DummyType>
     */
    private Collection $generatedData;
```

This removes the error when initialising the Collection. However then I get undefined property errors everywhere.

**Comment by LaravelFreelancerNL** [comment]
If the generated data can be 'anything' than [universal object crates](https://phpstan.org/config-reference#universal-object-crates) might be what you are looking for.

**Reply by jamespavett** [reply]
Sorry I'm not sure how that's relevant in this example, potentially I'm missing something.

Surely my collection does have a predefined structure. Just when it is initialised it contains no elements, and then DummyType elements will be added to it.

I didn't think this would be such an awkward use case...

**Comment by jamespavett** [comment]
Anyone have any ideas? @nunomaduro I know you said to raise it here, not sure if you have any thoughts.

**Comment by bastien-phi** [comment]
@jamespavett There was an issue raised in phpstan (https://github.com/phpstan/phpstan/issues/2683) for this.

Actually, there is no many ways to define your empty collection as an `\Illuminate\Support\Collection<int, DummyType>`

First is pretty simple but a bit verbose IMO : 
```php
public function __construct()
{
    /** @var array<int, DummyType> $generatedData */
    $generatedData = [];
    $this->generatedData = new Collection($generatedData);
}
```
You can also define te collection : 
```php
public function __construct()
{
    /** @var \Illuminate\Support\Collection<int, DummyType> $generatedData */
    $generatedData =  new Collection();
    $this->generatedData = $generatedData;
}
```

Alternatively, I've come to a kind of helper (not very pretty) to define an empty typed collection : 
```php
/**
 * @template TKey of array-key
 * @template TValue
 *
 * @param TKey $key
 * @param class-string<TValue> $value
 * @return \Illuminate\Support\Collection<TKey, TValue>
*/
function emptyCollection(int|string $key, string $value): Collection
{
    /** @var Collection<TKey, TValue> $collection */
    $collection = new Collection();

    return $collection;
}
```
You can then write your constructor with
```php
public function __construct()
{
    $this->generatedData = emptyCollection(0, DummyType::class);
}
```

I would be better to be able to write `emptyCollection('int', DummyType::class)` or `emptyCollection('string', DummyType::class)`  but I wasn't able to make it work.

If anyone has a better solution, I'll take it !

**Reply by bastien-phi** [reply]
For whose interested, I just released a package to help with this : https://github.com/Soyhuce/empty-collection

```php
/**
 * The generated combined data
 *
 * @var \Illuminate\Support\Collection<int, DummyType>
 */
private Collection $generatedData;

/**
 * Constructor.
 */
public function __construct()
{
    $this->generatedData = empty_collection('int', DummyType::class);
}
```

It supports, for exemple : 
```php
empty_collection('int', 'int'); // Illuminate\Support\Collection<int, int>
empty_collection('int', 'string'); // Illuminate\Support\Collection<int, string>
empty_collection('string', 'array<string, bool>'); // Illuminate\Support\Collection<string, array<string, bool>>
empty_collection('string', Item::class); // Illuminate\Support\Collection<int, App\Service\Item>
empty_eloquent_collection('int', User::class); // Illuminate\Database\Eloquent\Collection<int, App\Models\User>
empty_eloquent_collection('string', User::class); // Illuminate\Database\Eloquent\Collection<int, App\Models\User>
```

**Reply by szepeviktor** [reply]
@bastien-phi Stealing my idea?? 🤣 https://gist.github.com/szepeviktor/6003fb6e2a35c0e5f9eb29c96623e4b7

**Reply by bastien-phi** [reply]
:joy: No better way for now :man_shrugging:

**Comment by canvural** [comment, marked as answer]
This might be fixed in PHPStan. Can anyone try `dev-master` of PHPStan and check if it's solved?

**Reply by bastien-phi** [reply]
It works well. thanks @canvural !

Just be aware that the collection is untyped and no type checking will be performed.
```php
$items = new \Illuminate\Support\Collection();

$this->addItems($items);

$items
    ->each(function (Bar $bar, int $key) {
        $bar->execute($key);
    })
    ->each(function (Foo $foo, string $key) {
        $foo->run($key);
    });
```
is considered valid but should not ($items cannot be `Collection<int, Bar>` and `Collection<string, Foo>`)

In this case, you wille have to declare the collection type to make sure PHPStan understands correctly the type and raises the errors : 
```php
/** @var \Illuminate\Support\Collection<int, Bar> */
$items = new \Illuminate\Support\Collection();
```

**Comment by rudiedirkx** [comment]
This is how I do it:

    $checks = (new Check)->newCollection();

The correct collection is created, and Larastan knows what type is inside.

----

Only for `Eloquent\Collection` though.

**Reply by saulens22** [reply]
To anyone trying this - it's always possible to get `\Illuminate\Support\Collection` by doing this:

```php
$checks = (new Check)->newCollection()->toBase();
```
